botorch/models/utils/inducing_point_allocators.py

```diff
@@ -0,0 +1,388 @@
+        strategy of [moss2023ipa]_.
+
+        Args:
+            inputs: A (*batch_shape, n, d)-dim input data tensor.
+            covar_module: GPyTorch Module returning a LinearOperator kernel matrix.
+            num_inducing: The maximun number (m) of inducing points (m <= n).
+            input_batch_shape: The non-task-related batch shape.
+
+        Returns:
+            A (*batch_shape, m, d)-dim tensor of inducing point locations.
+        """
+
+        return self._allocate_inducing_points(
+            inputs,
+            covar_module,
+            num_inducing,
+            input_batch_shape,
+            ExpectedImprovementQualityFunction(self._model, self._maximize),
+        )
+
+
+def _pivoted_cholesky_init(
+    train_inputs: Tensor,
+    kernel_matrix: Union[Tensor, LinearOperator],
+    max_length: int,
+    quality_scores: Tensor,
+    epsilon: float = 1e-6,
+) -> Tensor:
+    r"""
+    A pivoted Cholesky initialization method for the inducing points,
+    originally proposed in [burt2020svgp]_ with the algorithm itself coming from
+    [chen2018dpp]_. Code is a PyTorch version from [chen2018dpp]_, based on
+    https://github.com/laming-chen/fast-map-dpp/blob/master/dpp.py but with a small
+    modification to allow the underlying DPP to be defined through its diversity-quality
+    decomposition,as discussed by [moss2023ipa]_. This method returns a greedy
+    approximation of the MAP estimate of the specified DPP, i.e. its returns a
+    set of points that are highly diverse (according to the provided kernel_matrix)
+    and have high quality (according to the provided quality_scores).
+
+    Args:
+        train_inputs: training inputs (of shape n x d)
+        kernel_matrix: kernel matrix on the training
+            inputs
+        max_length: number of inducing points to initialize
+        quality_scores: scores representing the quality of each candidate
+            input (of shape [n])
+        epsilon: numerical jitter for stability.
+
+    Returns:
+        max_length x d tensor of the training inputs corresponding to the top
+        max_length pivots of the training kernel matrix
+    """
+
+    # this is numerically equivalent to iteratively performing a pivoted cholesky
+    # while storing the diagonal pivots at each iteration
+    # TODO: use gpytorch's pivoted cholesky instead once that gets an exposed list
+    # TODO: ensure this works in batch mode, which it does not currently.
+
+    # todo test for shape of quality function
+
+    if quality_scores.shape[0] != train_inputs.shape[0]:
+        raise ValueError(
+            "_pivoted_cholesky_init requires a quality score for each of train_inputs"
+        )
+
+    item_size = kernel_matrix.shape[-2]
+    cis = torch.zeros(
+        (max_length, item_size), device=kernel_matrix.device, dtype=kernel_matrix.dtype
+    )
+    di2s = kernel_matrix.diag()
+    scores = di2s * (quality_scores**2)
```

Comment:
When any of the tests [here](https://github.com/pytorch/botorch/blob/d026e805f34bcffc9003720f1822e06f92ec08ab/test/sampling/pathwise/test_posterior_samplers.py#L32) is run on a GPU, this generates 
```
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```

Reply:
how do I fix this?

Reply:
You'll need to make sure the tensors here are on the same device; you could (and probably should) do this here with a `to()` call. But ideally this could already happen upstream where `_pivoted_cholesky_init` is called so that the inputs are already on the same device.

I think the reason this is happening here is that in `_allocate_inducing_points` the `covar_module` and `inputs` don't live on the device in the tests.

Reply:
so is it a quirk of the tests or an error in my code? Is it not fair to assume that `covar_module` and `inputs` are on device?

Reply:
The test seems to be passing on a GPU now.